### PR TITLE
add provides-extras to lockfile

### LINKED ISSUE #10953: Add `provides-extra` to the lockfile
### Summary

We write `requires-dist`, but not `provides-extra`, so it's not possible to determine the set of valid extras (since an extra could be empty, in which case, it wouldn't appear in `requires-dist`).

### PR DESCRIPTION
**Pull request by Gankra**
Fixes #10953

### CODE REVIEW

**Comment by Gankra** [comment]
I still need to add a snapshot test where this is non-empty.

File: crates/uv-resolver/src/lock/mod.rs [hunk 1]
```diff
@@ -1786,6 +1786,16 @@ impl Package {
                 .collect::<Result<_, _>>()
                 .map_err(LockErrorKind::RequirementRelativePath)?
         };
+        let provides_extras = if id.source.is_immutable() {
+            Vec::new()
```
Comment (charliermarsh) [inline crates/uv-resolver/src/lock/mod.rs:1790]: Something we need to consider: if we want to start using this to _enforce_ that a given extra exists (at install time, by reading from the lockfile), we won't be able to tell the difference between "empty" and "absent". So we either need to _not_ do that, or bump the lockfile version so we can tell the difference, or write these even when empty.

File: crates/uv-resolver/src/lock/mod.rs [hunk 2]
```diff
@@ -1786,6 +1786,16 @@ impl Package {
...
+        } else {
+            annotated_dist
+                .metadata
+                .as_ref()
+                .expect("metadata is present")
+                .provides_extras
+                .clone()
+        };
```
Comment (charliermarsh) [inline crates/uv-resolver/src/lock/mod.rs:1798]: I think you need to change `to_toml` to write these out. The impl is manual.

File: crates/uv-resolver/src/lock/mod.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv-resolver/src/lock/mod.rs:1790]: I'm not sure. @zanieb, do you have an opinion? Bumping the lockfile version means that older uv versions will reject newer lockfiles (but we can still read older lockfiles in newer uv versions).

Comment (Gankra) [reply crates/uv-resolver/src/lock/mod.rs:1790]: In addition to always writing this we'd also need to make this like... an `Option<Vec<ExtraName>>` to distinguish whether we found it in the lockfile, right?

Comment (charliermarsh) [reply crates/uv-resolver/src/lock/mod.rs:1790]: Yeah that’s right

Comment (Gankra) [reply crates/uv-resolver/src/lock/mod.rs]: Added the commit, I'm not sure we're gonna love how effectively `[package.metadata]` is no longer ever empty-and-therefore-omitable.

Comment (charliermarsh) [reply crates/uv-resolver/src/lock/mod.rs]: I was thinking that we'd continue to omit it for immutable sources, is that doable? It should massively reduce the diff.

Comment (Gankra) [reply crates/uv-resolver/src/lock/mod.rs]: It did remove a lot of them, yeah. (updated the commit)

Comment (zanieb) [reply crates/uv-resolver/src/lock/mod.rs]: Generally in favor of less breaking lockfile changes :) it's decent timing with 0.6 coming up, but from what I understand this doesn't justify the churn?

Comment (Gankra) [reply crates/uv-resolver/src/lock/mod.rs]: It's just to enable more validation on cli arguments, yeah. Not really worth a break.

Comment (charliermarsh) [reply crates/uv-resolver/src/lock/mod.rs]: Yeah I don't think we should bump for it. I do wonder if we should treat `requires-dist` the same way, though: always write it, even if empty?

I think users will likely be annoyed that the lockfile is churning more (even without a bump) in that you're now getting different results if you lock with a later version (even though it should be backwards-compatible), but I don't know how we avoid that.

**Comment by Gankra** [comment]
> CI / check system | python3.13 on windows x86-64 (pull_request) 

this task is extremely haunted on this PR and I wish I had any reason to think this PR breaks it...

**Comment by Gankra** [comment]
> Error: Cache folder path is retrieved for pip but doesn't exist on disk: c:\users\runneradmin\appdata\local\pip\cache. This likely indicates that there are no dependencies to cache. Consider removing the cache step if it is not needed.

god worst reason to die too

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
> god worst reason to die too

I think @zanieb fixed this on `main` so this branch might just be behind.
